### ISSUE TITLE
CrossCorrelLoss以及FID

### ISSUE THREAD

**Issue by holydick99** [Y-debug-sys/Diffusion-TS#73]
![image](https://github.com/user-attachments/assets/ccaa83dc-5127-4e3e-95a5-b68e4f838e33)
作者你好，我在使用的对应的FID以及CrossCorrelLoss后FID在1.4左右，但loss= 1.2441e-07，我对这样的loss值感到疑惑

**Comment by Y-debug-sys** [comment]
你好，这两个metric之间没什么直接联系吧，请问疑问是？

**Comment by holydick99** [comment]
> 你好，这两个metric之间没什么直接联系吧，请问疑问是？

你好，FID的表现较为正常，但是因为原论文中Correlational Score表现最好的数据集在0.004左右也就是10的-3次方，但是在我的数据集上cross_correl_loss达到了10的-7次方这个区间，所以我想问一下是否是因为函数引用错误导致，还是别的原因

**Comment by Y-debug-sys** [comment]
首先数据对比是归一化后的数据，其次这个值和数据集相关，谢谢。

**Comment by Y-debug-sys** [comment]
首先数据对比是归一化后的数据，其次这个值和数据集相关，谢谢。

**Comment by Y-debug-sys** [comment]
是的，放大dropout也可以。

**Comment by holydick99** [comment]
作者你好，下面是我的训练结果：好奇是什么导致t-sne的可视化出现这种情况

![image](https://github.com/user-attachments/assets/41cf6dd2-e789-4c4e-ba6d-a6163f60c5ce)
![image](https://github.com/user-attachments/assets/eba77233-defc-4065-8e2b-42e0cb80cc2d)
![image](https://github.com/user-attachments/assets/402c4b82-8152-478b-b2bb-2bc5d22c52cb)
![image](https://github.com/user-attachments/assets/82d27d32-f4fe-4f67-be64-e219d2b5a070)

**Comment by Y-debug-sys** [comment]
`epilepsy_ground_truth_178_train` 是归一化数据吗？

**Comment by holydick99** [comment]
> `epilepsy_ground_truth_178_train.npy` 是归一化数据吗？

下面是数据集加载代码
import os
import torch
import numpy as np
import pandas as pd

from scipy import io
from sklearn.preprocessing import MinMaxScaler
from torch.utils.data import Dataset
from Models.interpretable_diffusion.model_utils import normalize_to_neg_one_to_one, unnormalize_to_zero_to_one
from Utils.masking_utils import noise_mask


class EpilepsyDataset(Dataset):
   def __init__(
       self,
       name,
       data_root,
       window=64,
       proportion=0.8,
       save2npy=True,
       neg_one_to_one=True,
       seed=123,
       period='train',
       output_dir='./OUTPUT',
       predict_length=None,
       missing_ratio=None,
       style='separate',
       distribution='geometric',
       mean_mask_length=3
   ):
       super(EpilepsyDataset, self).__init__()
       assert period in ['train', 'test'], 'period must be train or test.'
       if period == 'train':
           assert ~(predict_length is not None or missing_ratio is not None), ''
       
       self.auto_norm = neg_one_to_one
       self.dir = os.path.join(output_dir, 'samples')
       os.makedirs(self.dir, exist_ok=True)
       self.pred_len, self.missing_ratio = predict_length, missing_ratio
       self.style, self.distribution, self.mean_mask_length = style, distribution, mean_mask_length

       self.rawdata, self.scaler = self.read_data(data_root)
       self.window, self.var_num = self.rawdata.shape[1], self.rawdata.shape[2]

       self.period, self.save2npy = period, save2npy
       self.samples = self.normalize(self.rawdata)
       self.sample_num = self.samples.shape[0]

       if period == 'test':
           if missing_ratio is not None:
               self.masking = self.mask_data(seed)
           elif predict_length is not None:
               masks = np.ones(self.samples.shape)
               masks[:, -predict_length:, :] = 0
               self.masking = masks.astype(bool)
           else:
               raise NotImplementedError()

   def normalize(self, sq):
       d = self.__normalize(sq.reshape(-1, self.var_num))
       data = d.reshape(-1, self.window, self.var_num)
       if self.save2npy:
           np.save(os.path.join(self.dir, f"epilepsy_ground_truth_{self.window}_{self.period}.npy"), sq)

           if self.auto_norm:
               np.save(os.path.join(self.dir, f"epilepsy_norm_truth_{self.window}_{self.period}.npy"), unnormalize_to_zero_to_one(data))
           else:
               np.save(os.path.join(self.dir, f"epilepsy_norm_truth_{self.window}_{self.period}.npy"), data)

       return data

   def __normalize(self, rawdata):
       data = self.scaler.transform(rawdata)
       if self.auto_norm:
           data = normalize_to_neg_one_to_one(data)
       return data

   def unnormalize(self, sq):
       d = self.__unnormalize(sq.reshape(-1, self.var_num))
       return d.reshape(-1, self.window, self.var_num)

   def __unnormalize(self, data):
       if self.auto_norm:
           data = unnormalize_to_zero_to_one(data)
       x = data
       return self.scaler.inverse_transform(x)
   
   def mask_data(self, seed=2023):
       masks = np.ones_like(self.samples)
       # Store the state of the RNG to restore later.
       st0 = np.random.get_state()
       np.random.seed(seed)

       for idx in range(self.samples.shape[0]):
           x = self.samples[idx, :, :]  # (seq_length, feat_dim) array
           mask = noise_mask(x, self.missing_ratio, self.mean_mask_length, self.style,
                             self.distribution)  # (seq_length, feat_dim) boolean array
           masks[idx, :, :] = mask

       if self.save2npy:
           np.save(os.path.join(self.dir, f"epilepsy_masking_{self.window}.npy"), masks)

       # Restore RNG.
       np.random.set_state(st0)
       return masks.astype(bool)

   def __getitem__(self, ind):
       if self.period == 'test':
           x = self.samples[ind, :, :]  # (seq_length, feat_dim) array
           m = self.masking[ind, :, :]  # (seq_length, feat_dim) boolean array
           return torch.from_numpy(x).float(), torch.from_numpy(m)
       x = self.samples[ind, :, :]  # (seq_length, feat_dim) array
       return torch.from_numpy(x).float()

   def __len__(self):
       return self.sample_num

   @staticmethod
   def read_data(filepath, name=''):
       """Reads a single .csv
       """
#         可选用不同置信度的数据集训练
       data = io.loadmat(filepath + '/pseudo_labels_epilepsy.mat')['Epilepsy']
       
#         data = io.loadmat(filepath + '/pseudo_labels_epilepsy0.9.mat')['Epilepsy']
       
#         data = io.loadmat(filepath + '/pseudo_labels_epilepsy0.95.mat')['Epilepsy']
       
#         data = io.loadmat(filepath + '/pseudo_labels_epilepsy0.99.mat')['Epilepsy']
       
       scaler = MinMaxScaler()
       scaler = scaler.fit(data.reshape(-1, 1))
       return data.reshape(data.shape[0], -1, 1), scaler
       
       这是我构建数据集的代码，根据代码来看，是的

**Comment by holydick99** [comment]
根据我加载的代码来看，是的
下面是数据集加载代码
’‘’
import os
import torch
import numpy as np
import pandas as pd

from scipy import io
from sklearn.preprocessing import MinMaxScaler
from torch.utils.data import Dataset
from Models.interpretable_diffusion.model_utils import normalize_to_neg_one_to_one, unnormalize_to_zero_to_one
from Utils.masking_utils import noise_mask


class EpilepsyDataset(Dataset):
   def __init__(
       self,
       name,
       data_root,
       window=64,
       proportion=0.8,
       save2npy=True,
       neg_one_to_one=True,
       seed=123,
       period='train',
       output_dir='./OUTPUT',
       predict_length=None,
       missing_ratio=None,
       style='separate',
       distribution='geometric',
       mean_mask_length=3
   ):
       super(EpilepsyDataset, self).__init__()
       assert period in ['train', 'test'], 'period must be train or test.'
       if period == 'train':
           assert ~(predict_length is not None or missing_ratio is not None), ''
       
       self.auto_norm = neg_one_to_one
       self.dir = os.path.join(output_dir, 'samples')
       os.makedirs(self.dir, exist_ok=True)
       self.pred_len, self.missing_ratio = predict_length, missing_ratio
       self.style, self.distribution, self.mean_mask_length = style, distribution, mean_mask_length

       self.rawdata, self.scaler = self.read_data(data_root)
       self.window, self.var_num = self.rawdata.shape[1], self.rawdata.shape[2]

       self.period, self.save2npy = period, save2npy
       self.samples = self.normalize(self.rawdata)
       self.sample_num = self.samples.shape[0]

       if period == 'test':
           if missing_ratio is not None:
               self.masking = self.mask_data(seed)
           elif predict_length is not None:
               masks = np.ones(self.samples.shape)
               masks[:, -predict_length:, :] = 0
               self.masking = masks.astype(bool)
           else:
               raise NotImplementedError()

   def normalize(self, sq):
       d = self.__normalize(sq.reshape(-1, self.var_num))
       data = d.reshape(-1, self.window, self.var_num)
       if self.save2npy:
           np.save(os.path.join(self.dir, f"epilepsy_ground_truth_{self.window}_{self.period}.npy"), sq)

           if self.auto_norm:
               np.save(os.path.join(self.dir, f"epilepsy_norm_truth_{self.window}_{self.period}.npy"), unnormalize_to_zero_to_one(data))
           else:
               np.save(os.path.join(self.dir, f"epilepsy_norm_truth_{self.window}_{self.period}.npy"), data)

       return data

   def __normalize(self, rawdata):
       data = self.scaler.transform(rawdata)
       if self.auto_norm:
           data = normalize_to_neg_one_to_one(data)
       return data

   def unnormalize(self, sq):
       d = self.__unnormalize(sq.reshape(-1, self.var_num))
       return d.reshape(-1, self.window, self.var_num)

   def __unnormalize(self, data):
       if self.auto_norm:
           data = unnormalize_to_zero_to_one(data)
       x = data
       return self.scaler.inverse_transform(x)
   
   def mask_data(self, seed=2023):
       masks = np.ones_like(self.samples)
       # Store the state of the RNG to restore later.
       st0 = np.random.get_state()
       np.random.seed(seed)

       for idx in range(self.samples.shape[0]):
           x = self.samples[idx, :, :]  # (seq_length, feat_dim) array
           mask = noise_mask(x, self.missing_ratio, self.mean_mask_length, self.style,
                             self.distribution)  # (seq_length, feat_dim) boolean array
           masks[idx, :, :] = mask

       if self.save2npy:
           np.save(os.path.join(self.dir, f"epilepsy_masking_{self.window}.npy"), masks)

       # Restore RNG.
       np.random.set_state(st0)
       return masks.astype(bool)

   def __getitem__(self, ind):
       if self.period == 'test':
           x = self.samples[ind, :, :]  # (seq_length, feat_dim) array
           m = self.masking[ind, :, :]  # (seq_length, feat_dim) boolean array
           return torch.from_numpy(x).float(), torch.from_numpy(m)
       x = self.samples[ind, :, :]  # (seq_length, feat_dim) array
       return torch.from_numpy(x).float()

   def __len__(self):
       return self.sample_num

   @staticmethod
   def read_data(filepath, name=''):
       """Reads a single .csv
       """
#         可选用不同置信度的数据集训练
       data = io.loadmat(filepath + '/pseudo_labels_epilepsy.mat')['Epilepsy']
       
#         data = io.loadmat(filepath + '/pseudo_labels_epilepsy0.9.mat')['Epilepsy']
       
#         data = io.loadmat(filepath + '/pseudo_labels_epilepsy0.95.mat')['Epilepsy']
       
#         data = io.loadmat(filepath + '/pseudo_labels_epilepsy0.99.mat')['Epilepsy']
       
       scaler = MinMaxScaler()
       scaler = scaler.fit(data.reshape(-1, 1))
       return data.reshape(data.shape[0], -1, 1), scaler
       ‘’‘

**Comment by Y-debug-sys** [comment]
![image](https://github.com/user-attachments/assets/2a224727-01b4-40a0-8bd1-e78bc7419f22)
